File: library-solana/poc/programs/tests/init_handle.ts
import * as anchor from "@coral-xyz/anchor";
import { Keypair } from "@solana/web3.js";
import { assert, expect } from "chai";
import {
  authority,
  derivePermissionList,
  ensureConfigInitialized,
  makeHandle,
  program,
  provider,
} from "./utils";

describe("acl :: init_handle", () => {
  before(ensureConfigInitialized);

  it("creates a permission_list PDA owned by the program with the given handle and an empty subject list", async () => {
    const handle = makeHandle(10);
    const initialKey = Keypair.generate().publicKey;
    const outputIndex = new anchor.BN(0);
    const permissionList = derivePermissionList(initialKey, outputIndex);

    await program.methods
      .initHandle(handle, initialKey, outputIndex)
      .accountsPartial({
        payer: provider.wallet.publicKey,
        authority: authority.publicKey,
        permissionList,
      })
      .signers([authority])
      .rpc();

    const acct = await program.account.handlerPermissions.fetch(permissionList);
    assert.deepEqual(Array.from(acct.handle), handle);
    assert.equal(acct.subjectCount, 0);

    const info = await provider.connection.getAccountInfo(permissionList);
    assert.isNotNull(info);
    assert.ok(info!.owner.equals(program.programId));
    // INIT_SPACE = handle(32) + state(1) + allowed_accounts(16*32=512)
    //              + subject_count(1) + bump(1) = 547
    // + 8-byte discriminator = 555
    assert.equal(info!.data.length, 555);
  });

  it("is idempotent when called twice with the same (initial_key, output_index, handle)", async () => {
    const handle = makeHandle(11);
    const initialKey = Keypair.generate().publicKey;
    const outputIndex = new anchor.BN(1);
    const permissionList = derivePermissionList(initialKey, outputIndex);

    for (let i = 0; i < 2; i++) {
      await program.methods
        .initHandle(handle, initialKey, outputIndex)
        .accountsPartial({
          payer: provider.wallet.publicKey,
          authority: authority.publicKey,
          permissionList,
        })
        .signers([authority])
        .rpc();
    }

    const acct = await program.account.handlerPermissions.fetch(permissionList);
    assert.deepEqual(Array.from(acct.handle), handle);
  });

  it("reverts with HandleMismatch when an already-bound PDA is re-bound to a different handle", async () => {
    const handle = makeHandle(12);
    const initialKey = Keypair.generate().publicKey;
    const outputIndex = new anchor.BN(2);
    const permissionList = derivePermissionList(initialKey, outputIndex);

    await program.methods
      .initHandle(handle, initialKey, outputIndex)
      .accountsPartial({
        payer: provider.wallet.publicKey,
        authority: authority.publicKey,
        permissionList,
      })
      .signers([authority])
      .rpc();

    try {
      const wrongHandle = handle.map((b) => b ^ 0xff);
      await program.methods
        .initHandle(wrongHandle, initialKey, outputIndex)
        .accountsPartial({
          payer: provider.wallet.publicKey,
          authority: authority.publicKey,
          permissionList,
        })
        .signers([authority])
        .rpc();
      assert.fail("expected HandleMismatch");
    } catch (err: any) {
      const code = err?.error?.errorCode?.code ?? "";
      const msg = code + " " + (err?.message ?? "");
      expect(msg).to.match(/HandleMismatch/);
    }
  });

  it("reverts with UnauthorizedAccess when the authority signer is not the fhe authority", async () => {
    const handle = makeHandle(13);
    const initialKey = Keypair.generate().publicKey;
    const outputIndex = new anchor.BN(3);
    const permissionList = derivePermissionList(initialKey, outputIndex);
    const impostor = Keypair.generate();

    try {
      await program.methods
        .initHandle(handle, initialKey, outputIndex)
        .accountsPartial({
          payer: provider.wallet.publicKey,
          authority: impostor.publicKey,
          permissionList,
        })
        .signers([impostor])
        .rpc();
      assert.fail("expected UnauthorizedAccess");
    } catch (err: any) {
      const code = err?.error?.errorCode?.code ?? "";
      const msg = code + " " + (err?.message ?? "");
      expect(msg).to.match(/UnauthorizedAccess/);
    }
  });
});

File: library-solana/poc/programs/programs/acl/src/instructions/init_handle.rs
use anchor_lang::prelude::*;

use crate::{
    constants, error::AclError, event::NewHandle, state::{Config, DISCRIMINATOR, HandlerPermissions, HandlerState}, types::Handle
};

#[event_cpi]
#[derive(Accounts)]
#[instruction(handle: Handle, initial_key: Pubkey, output_index: u128)]
pub struct InitHandle<'info> {
    #[account(mut)]
    pub payer: Signer<'info>,
    pub authority: Signer<'info>,

    #[account(
        init_if_needed,
        payer = payer,
        space = DISCRIMINATOR + HandlerPermissions::INIT_SPACE,
        seeds = [
            constants::PERMISSION_LIST, 
            initial_key.as_ref(), 
            &output_index.to_le_bytes()
        ],
        bump
    )]
    pub permission_list: Account<'info, HandlerPermissions>,

    #[account(
        seeds = [constants::ACL_CONFIG],
        bump = acl_config.bump
    )]
    pub acl_config: Account<'info, Config>,

    pub system_program: Program<'info, System>,
}

pub fn init_handle(ctx: Context<InitHandle>, handle: Handle, initial_key: Pubkey, output_index: u128) -> Result<()> {
    let config = &ctx.accounts.acl_config;
    let authority = &ctx.accounts.authority;
    let permission_list = &mut ctx.accounts.permission_list;
    require!(
        config.authorize(authority.key),
        AclError::UnauthorizedAccess
    );
    if permission_list.state == HandlerState::Bound {
        require!(permission_list.handle == handle, AclError::HandleMismatch);
        return Ok(());
    }
    permission_list.handle = handle;
    permission_list.state = HandlerState::Bound;
    permission_list.bump = ctx.bumps.permission_list;
    emit_cpi!(NewHandle {
        handle,
        initial_key,
        output_index,
        config_key: config.key()
    });

    Ok(())
}
